runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/runtime/jobs/Job.java

```diff
@@ -194,6 +196,82 @@ protected IStatus run(IProgressMonitor monitor) {
 		};
 	}
 
+	/**
+	 * Performs a one shot background job execution returning a value as an
+	 * syncronous computation to the caller with the follwoing side-effects:
+	 * <ul>
+	 * <li>If the {@link CompletableFuture} is canceled, best effort is made to also
+	 * cancel the job</li>
+	 * <li>If the backing {@link Job} is canceled, best effort it made to also
+	 * cancel the {@link CompletableFuture}</li>
+	 * </ul>
+	 *
+	 * @since 3.14
+	 */
+	public static CompletableFuture<?> execute(String name, IJobRunnable runnable) {
```

Comment:
Why a new IJobRunnable type instead of a `Consumer<IProgressMonitor>`?
Also the return type should be `CompletableFuture<Void>`.
Shouldn't we call it `runAsync()` like CompletableFuture API? or `schedule`? The current `name` fails at carrying the async nature IMO.

Reply:
> Why a new IJobRunnable type instead of a Consumer<IProgressMonitor>?

Mostly to allow for checked Exceptions.

> Also the return type should be CompletableFuture<Void>

What is the advantage of this?

> Shouldn't we call it runAsync() like CompletableFuture API? or schedule? The current name fails at carrying the async nature IMO.

For me the primary think here is to execute something but other names would be fine.

Reply:
> Mostly to allow for checked Exceptions.

CompletableFuture.get() does not throw an checked exception. and ExecutorService.submit does not accept a checked exception.

Reply:
> CompletableFuture.get() does not throw an checked exception

[Javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html) says different:

> In case of exceptional completion with a CompletionException, methods [get()](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#get--) and [get(long, TimeUnit)](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#get-long-java.util.concurrent.TimeUnit-) throw an [ExecutionException](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutionException.html) with the same cause as held in the corresponding CompletionException.

>  ExecutorService.submit does not accept a checked exception.

No idea how this relates here?

Beside what other API does, it is here for the user to not being forced to wrap checked exceptions into runtime exceptions, so the true exception can be accessed for example in 
- https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#exceptionally-java.util.function.Function-
- https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#handle-java.util.function.BiFunction-
- ... and its async variants ...

Reply:
CompletableFuture is meant to be used together with an Executor. All its async methods are defined to "are performed using the ForkJoinPool.commonPool()" - not what i expect them to do when generated by a eclipse. 
CancellationException/OperationCanceledException may behave other then the handling of exception handling defined in the vice versa API .
CompletableFuture .cancel(boolean mayInterruptIfRunning) will probably behave other then defined.
I am not convinced that it is a good idea to mix java.util.concurrent with eclipse stuff. Both have their  pros and cons, but it will be hard or impossible to get them together so that they fullfill both contracts for all edge cases.

Reply:
> CompletableFuture is meant to be used together with an Executor.

Only the async methods allow to specify a separate executor, but this has nothing to do with exception handling or how the _initial value_ is computed (and where exceptions are happening), because all later stages only operate on a _result_ and this might be processed where ever (even in another job if desired).

> not what i expect them to do when generated by a eclipse

Eclipse uses the common poll al time long if it is using streams, so I don't really see how this relates here, this is all only about computing the inital result, the rest is plain java and is not affected at all by where it runs...

> CompletableFuture .cancel(boolean mayInterruptIfRunning) will probably behave other then defined.

Please explain or give concrete example "probably" is nothing one can act upon.

> but it will be hard or impossible to get them together so that they fullfill both contracts for all edge cases.

There is essentially no need to fulfill *all* contracts (beside "Eclipse" and "java.util.concurrent" are not mutual exclusive in any way).

Reply:
See for example https://github.com/eclipse-platform/eclipse.platform/issues/294 for problems with commonPool() in eclipse. Eclipse normally does NOT use stream().parallel.
If the job throws CancellationException or OperationCanceledException what should happen? what should isCancelled() return? what is the IStatus?
mayInterruptIfRunning: see https://github.com/eclipse-platform/eclipse.platform/pull/326#discussion_r1097353063